File: frontend/js/views/margin_analysis.js
// Income-statement margin waterfall — gross / operating / pre-tax / net profit
// and margins, via /calc/margin-analysis. Live as you type.

import { api } from '../api.js';
import { applyUiI18n, t } from '../i18n.js';
import { currentViewToken, viewIsCurrent } from '../app.js';
import { showToast } from '../toast.js';
import { debounce } from '../util.js';

const money = (n) => '$' + Number(n).toLocaleString(undefined, { minimumFractionDigits: 2, maximumFractionDigits: 2 });
const pct = (n) => (n == null ? '—' : Number(n).toLocaleString(undefined, { maximumFractionDigits: 2 }) + '%');

export async function renderMarginAnalysis(mount, _appState) {
    const tok = currentViewToken();
    if (!viewIsCurrent(tok)) return;
    mount.innerHTML = `
        <h1 class="view-title"><span data-i18n="view.margin.h1.title">// MARGIN ANALYSIS</span></h1>
        <p class="muted small" data-i18n="view.margin.hint.intro">
            The income-statement waterfall: revenue down to net income, with the margin (percent of
            revenue) at each level. Gross strips out COGS, operating subtracts overhead, pre-tax
            removes interest, and net is after tax. Updates as you type.
        </p>
        <div class="lpv-split">
        <div class="chart-panel">
            <h2 data-i18n="view.margin.h2.inputs">The income statement</h2>
            <form id="margin-form" class="inline-form">
                <label><span data-i18n="view.margin.label.revenue">Revenue ($)</span>
                    <input type="number" step="0.01" min="0" name="revenue_usd" value="1000" required></label>
                <label><span data-i18n="view.margin.label.cogs">COGS ($)</span>
                    <input type="number" step="0.01" min="0" name="cogs_usd" value="600" required></label>
                <label><span data-i18n="view.margin.label.opex">Operating expenses ($)</span>
                    <input type="number" step="0.01" min="0" name="operating_expenses_usd" value="200"></label>
                <label><span data-i18n="view.margin.label.interest">Interest expense ($)</span>
                    <input type="number" step="0.01" min="0" name="interest_expense_usd" value="50"></label>
                <label><span data-i18n="view.margin.label.tax">Tax rate (%)</span>
                    <input type="number" step="0.1" min="0" name="tax_rate_pct" value="21"></label>
            </form>
        </div>
        <div id="margin-result" class="lpv-preview"></div>
        </div>
    `;
    applyUiI18n(mount);

    const form = mount.querySelector('#margin-form');
    const generate = async () => {
        const fd = new FormData(form);
        const body = {
            revenue_usd: Number(fd.get('revenue_usd')) || 0,
            cogs_usd: Number(fd.get('cogs_usd')) || 0,
            operating_expenses_usd: Number(fd.get('operating_expenses_usd')) || 0,
            interest_expense_usd: Number(fd.get('interest_expense_usd')) || 0,
            tax_rate_pct: Number(fd.get('tax_rate_pct')) || 0,
        };
        try {
            const r = await api.calcMarginAnalysis(body);
            if (!viewIsCurrent(tok)) return;
            renderResult(mount, r);
        } catch (err) {
            showToast(err.message || t('view.margin.toast.error'), { level: 'error' });
        }
    };
    form.addEventListener('input', debounce(generate, 250));
    form.addEventListener('submit', (e) => { e.preventDefault(); generate(); });
    generate();
}

function renderResult(mount, r) {
    const el = mount.querySelector('#margin-result');
    const netCls = r.net_income_usd >= 0 ? 'pos' : 'neg';
    el.innerHTML = `
        <div class="chart-panel">
            <h2 data-i18n="view.margin.h2.result">The margins</h2>
            <div class="cards">
                <div class="card"><div class="label" data-i18n="view.margin.card.gross">Gross margin</div>
                    <div class="value">${pct(r.gross_margin_pct)}</div></div>
                <div class="card"><div class="label" data-i18n="view.margin.card.operating">Operating margin</div>
                    <div class="value">${pct(r.operating_margin_pct)}</div></div>
                <div class="card ${netCls}"><div class="label" data-i18n="view.margin.card.net">Net margin</div>
                    <div class="value ${netCls}">${pct(r.net_margin_pct)}</div></div>
            </div>
            <table class="data-table">
                <thead><tr><th data-i18n="view.margin.col.line">Line</th><th data-i18n="view.margin.col.amount">Amount</th><th data-i18n="view.margin.col.margin">Margin</th></tr></thead>
                <tbody>
                    <tr><td data-i18n="view.margin.row.gross">Gross profit</td><td>${money(r.gross_profit_usd)}</td><td>${pct(r.gross_margin_pct)}</td></tr>
                    <tr><td data-i18n="view.margin.row.operating">Operating income</td><td>${money(r.operating_income_usd)}</td><td>${pct(r.operating_margin_pct)}</td></tr>
                    <tr><td data-i18n="view.margin.row.pretax">Pre-tax income</td><td>${money(r.pretax_income_usd)}</td><td>${pct(r.pretax_margin_pct)}</td></tr>
                    <tr><td data-i18n="view.margin.row.tax">Tax</td><td>${money(r.tax_usd)}</td><td>—</td></tr>
                    <tr class="emph ${netCls}"><td data-i18n="view.margin.row.net">Net income</td><td>${money(r.net_income_usd)}</td><td>${pct(r.net_margin_pct)}</td></tr>
                </tbody>
            </table>
        </div>
    `;
    applyUiI18n(el);
}

File: crates/traderview-core/src/margin_analysis.rs
//! Income-statement margin waterfall — gross, operating, pre-tax, and net
//! profit at each step down the P&L, with the margin (percent of revenue) at
//! every level.
//!
//! ```text
//! gross profit     = revenue − COGS
//! operating income = gross profit − operating expenses   (EBIT)
//! pre-tax income   = operating income − interest
//! net income       = pre-tax − tax (tax on positive pre-tax only)
//! ```

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Deserialize)]
pub struct MarginInput {
    pub revenue_usd: f64,
    pub cogs_usd: f64,
    #[serde(default)]
    pub operating_expenses_usd: f64,
    #[serde(default)]
    pub interest_expense_usd: f64,
    /// Tax rate on positive pre-tax income, percent.
    #[serde(default)]
    pub tax_rate_pct: f64,
}

#[derive(Debug, Clone, Serialize, PartialEq)]
pub struct MarginResult {
    pub gross_profit_usd: f64,
    pub operating_income_usd: f64,
    pub pretax_income_usd: f64,
    pub tax_usd: f64,
    pub net_income_usd: f64,
    /// Each as a percent of revenue; `None` if revenue ≤ 0.
    pub gross_margin_pct: Option<f64>,
    pub operating_margin_pct: Option<f64>,
    pub pretax_margin_pct: Option<f64>,
    pub net_margin_pct: Option<f64>,
}

pub fn analyze(input: &MarginInput) -> MarginResult {
    let gross = input.revenue_usd - input.cogs_usd;
    let operating = gross - input.operating_expenses_usd;
    let pretax = operating - input.interest_expense_usd;
    let tax = if pretax > 0.0 {
        pretax * input.tax_rate_pct / 100.0
    } else {
        0.0
    };
    let net = pretax - tax;

    let margin = |x: f64| {
        if input.revenue_usd > 0.0 {
            Some(x / input.revenue_usd * 100.0)
        } else {
            None
        }
    };

    MarginResult {
        gross_profit_usd: gross,
        operating_income_usd: operating,
        pretax_income_usd: pretax,
        tax_usd: tax,
        net_income_usd: net,
        gross_margin_pct: margin(gross),
        operating_margin_pct: margin(operating),
        pretax_margin_pct: margin(pretax),
        net_margin_pct: margin(net),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn close(a: f64, b: f64) -> bool {
        (a - b).abs() < 1e-4
    }

    fn base() -> MarginInput {
        MarginInput {
            revenue_usd: 1000.0,
            cogs_usd: 600.0,
            operating_expenses_usd: 200.0,
            interest_expense_usd: 50.0,
            tax_rate_pct: 21.0,
        }
    }

    #[test]
    fn gross() {
        let r = analyze(&base());
        assert!(close(r.gross_profit_usd, 400.0));
        assert!(close(r.gross_margin_pct.unwrap(), 40.0));
    }

    #[test]
    fn operating() {
        let r = analyze(&base());
        assert!(close(r.operating_income_usd, 200.0));
        assert!(close(r.operating_margin_pct.unwrap(), 20.0));
    }

    #[test]
    fn pretax() {
        let r = analyze(&base());
        assert!(close(r.pretax_income_usd, 150.0));
        assert!(close(r.pretax_margin_pct.unwrap(), 15.0));
    }

    #[test]
    fn tax_on_pretax() {
        assert!(close(analyze(&base()).tax_usd, 31.5));
    }

    #[test]
    fn net() {
        let r = analyze(&base());
        assert!(close(r.net_income_usd, 118.5));
        assert!(close(r.net_margin_pct.unwrap(), 11.85));
    }

    #[test]
    fn margins_narrow_down_the_waterfall() {
        let r = analyze(&base());
        assert!(
            r.gross_margin_pct.unwrap()
                > r.operating_margin_pct.unwrap()
                && r.operating_margin_pct.unwrap() > r.net_margin_pct.unwrap()
        );
    }

    #[test]
    fn loss_has_no_tax() {
        // Heavy opex drives a pre-tax loss → no tax charged.
        let r = analyze(&MarginInput {
            operating_expenses_usd: 500.0,
            ..base()
        });
        assert!(r.pretax_income_usd < 0.0);
        assert!(close(r.tax_usd, 0.0));
        // Net loss equals pre-tax loss (no tax benefit modeled).
        assert!(close(r.net_income_usd, r.pretax_income_usd));
    }

    #[test]
    fn zero_revenue_guards_margins() {
        let r = analyze(&MarginInput {
            revenue_usd: 0.0,
            ..base()
        });
        assert!(r.gross_margin_pct.is_none());
        assert!(r.net_margin_pct.is_none());
    }
}
